### ISSUE TITLE
Feature request: E-Stim control using audio

### ISSUE THREAD

**Issue by teepox** [Yoooi0/MultiFunPlayer#65]
Support for E-Stim could be implemented using a separate audio file (MP3/WAV) for input, and selectable audio device for output. In this model MultiFunPlayer would act as an audio player, and sync play/pause for both input video and E-Stim audio. Audio from the video file and audio for E-Stim would be routed to separate audio devices for playback.

A more basic model would be to use an external audio player for the E-Stim audio. For a more advanced implementation, MultiFunPlayer could generate an audio signal from a funscript. Something similar has already been implemented in ScriptPlayer v1.1.1.

**Comment by diglet48** [comment]
I made a tool to dynamically generate estim audio. The audio is generated based on (currently) 2-axis input, it accepts tcode commands over websockets so it integrates well with MFP. Saved me a lot of time by not having to implement video synchronization myself.

https://github.com/diglet48/restim

If the project gets a bit more users perhaps we can add a device preset in MFP, but no other changes to MFP are necessary.

**Comment by Yoooi0** [comment]
Nice! I guess it should work with any player that can send TCode via a websocket?

I would probably still add some version of estim support to MFP, just need to find a good resource on what the actual audio output should be and what are expected configurable parameters.

**Comment by edger477** [comment]
> MultiFunPlayer will never directly handle video or audio, only scripts.
Why? Especially if you will consider adding audio output, why not have a source that just takes already existing mp3 and doesn't do anything with scripts. Half of my library are videos without scripts, and I was sent here from milovana when I saw that it syncs with HereSphere... so I wanted to try diglet's restim (will probably still try it) but I have to keep also using ScriptPlayer for syncing with videos that have dedicated mp3 estim. Actually I wanted to pull code to see if I could contribute something but need to know why you are categorically against it.

**Comment by Yoooi0** [comment]
I just don't want to have a full video/audio player to support when almost everything can be done via mpv. 
There is also no easy place to fit that in that I can think of. Generating audio could be easily done in an output target which runs in its own thread while decoding+playing+synchronizing audio/video internally is a different beast.

Tbh I dont even know how estim mp3 are used. Are they synced to the movement in the video? Do you have to output them to separate audio output?
Maybe this could be done via MFP plugin? Which opens and synchronizes a second mpv window just for the estim mp3.

**Comment by edger477** [comment]
Oh, ok let me explain. Many people use estim instead of strokers. diglet's restim does on-the-fly conversion of funscript into estim signals that provide similar sensations, but it is much more advanced since it allows tuning how the processing is done on the fly. All that is great for "standard" videos that are often POV, with funscript, but we also have many creators that create videos that have its own estim, sometimes also funscript, but estim they produce can contain things like strong tremolo that is intended to be a challenging section (edging or almost painful), or to cause effects that are hard to accomplish with funscript, sometimes also producing a dual-channel (so no triphase type of signal that we generate from funscript but something intended for different experience).

The ScriptPlayer, although it can also play video in itself, does similar job as MFP in that it can use either local player, or sync to HereSphere, MPC etc, and in that case it is only playing either .funscript, or a .mp3/.wav file (or both). You can define outputs, there is "estim output" (sound card) to which mp3 file with same name as video is automatically played in sync with video (whatever is your player of choice). So I mainly use it to sync mp3 files that contain estim signals for videos that I play in HereSphere.

Now, many of these mp3 files are converts from funscript, some of them I even did in my fork of existing funscript->mp3 converter (https://edger477.github.io/funstim/funstim.html), but I saw what diglet did with realtime conversion and I want to try that. ScriptPlayer has rudimentary ability to convert funscripts to audio device output in realtime, but I do not use it since the output is choppy. I have a feeling I will like the diglet's plug in and probably we can make a plugin in same way that would play mp3 in sync (same what restim does now funscript->audio card, it would just be mp3->audio card, so simpler because we don't need to do signal generation, we just take input buffer and pass to output). That plugin would completely fill the space that is now filled by ScriptPlayer since MFP+restim will be handling funscript -> estim audio (feature that now does not even work properly in ScriptPlayer), and then I guess we actually need to create a plugin that would do mp3->estim audio.

So, if I were to try to make a MFP plugin for syncplay of mp3 files to configurable audio device myself (mp3 would be loaded by name or convention based on currently played video from whatever is the main player, and synced to same timestamp), what existing plugin should I review to get started faster?

**Comment by goranobradovic** [comment]
After looking through the app, it seems to me like plugin that would be both input and output at same time might not fit into concept of the MFP, unless outputs can receive information about the (video) file that is being played in input source? Most of these plugins receive decoded funscript, but would it be possible to have output that would work when no funscript exists?

**Comment by Yoooi0** [comment]
By plugins I mean actual C# plugins. You put your plugin .cs file into the `Plugins` directory and it gets compiled at runtime. Here is the "api": https://github.com/Yoooi0/MultiFunPlayer/blob/master/Source/MultiFunPlayer/Plugin/PluginBase.cs
You extend either `AsyncPluginBase` or `SyncPluginBase` and you can listen/generate all the internal MFP events. You could listen to a few media messages and implement a simple mp3 player with something like NAudio.

> So, if I were to try to make a MFP plugin for syncplay of mp3 files to configurable audio device myself (mp3 would be loaded by name or convention based on currently played video from whatever is the main player, and synced to same timestamp), what existing plugin should I review to get started faster?

To have this native in MFP basically everything below the tab with the play/pause button (script heatmap and below) would have to be abstracted into a separate "container".
So from here:
https://github.com/Yoooi0/MultiFunPlayer/blob/362d0e3629e1846f8a4ba0b26af2d939f26f3717/Source/MultiFunPlayer/RootViewModel.cs#L15-L17
and here:
https://github.com/Yoooi0/MultiFunPlayer/blob/362d0e3629e1846f8a4ba0b26af2d939f26f3717/Source/MultiFunPlayer/RootView.xaml#L87-L88
The script stuff receives media messages, loads corresponding funscripts, and then sends data via one of the outputs.
Estim then would be implemented the same way, it would recieve media messages, load corresponding mp3 files, and then output to selected audio output.

Basically all that would have to be put into some sort of tab control, first tab would be for "scripts" second tab would be for "estim". "Estim" would probably also require a custom seek bar with audio waveform preview.
Same sort of tab control like for video players, where you can also enable/disable them so you dont have unnecessary threads running.

**Comment by Kelvinatorva** [comment]
I would like to echo diglet's and edger477's comments and requests for an audio input / output for estim mp3 files.  Diglet has done a great job in addressing one of estim's basic challenges: "How to make a one-size-fits-all estim signal feel good / comfortable / pleasurable ("painful in a good way), given the differences between estim box, electrodes and placement, and body chemistry (https://github.com/diglet48/restim  , refer to the github restim wiki for a more detailed explanation).  The restim individual calibration sequence has made a world of difference in improving the sensations and enjoyment of synchronized cock hero and estim videos.  It would be great if multifunplayer users could play some of the classic estim made-from-scratch mp3 files by digitalparkinglot, numberonefan, lolo2, and others along with their accompanying videos.
FYI - I have enjoyed MFP as an "orchestra  conductor" for my video estim pleasure.  There is a brief MFP installation and setup guide in the restim wiki aimed at helping the casual user use MFP.
Thanks for your efforts in developing and improving MFP.  Who knows, maybe one day I'll get a non audio device to play!  K

**Comment by Yoooi0** [comment]
Added basic PoC audio output, let me know if/how usable it is with estim:
https://github.com/Yoooi0/MultiFunPlayer/actions/runs/4772771530

**Comment by qwksilver** [comment]
I am confused at the current state of this issue and where and how to use the function if available.  There is a lot of talk in this thread about different ideas around implementation.  The most basic I can think of would be to have a small add on, or internal, audio player that is kept in time sync with the video.  Adjustment to +/- x ms to maximize the sync, and select output device, and desired output volume.  

There are so many ways to make this file type out there that automation around making estim tracks on the fly feels like a stretch goal, or better yet someone write it as a separate device that mfp can just send a sync signal and file info to.

**Comment by Yoooi0** [comment]
Like I said playing audio files directly in MFP is I think out of scope at the moment. It just adds too many dependencies.

There is a branch with simple implementation of audio generation from funscripts. It generates sine/square/saw/triangle waves with frequency/amplitude/balance values mapped from the funscript values. So you can have L0 drive the frequency from while R0 drives the amplitude.
But I never got any feedback so it was never released. I only have coyote 3 which I can add support for directly without audio generation since it has a simple bluetooth api. From the looks of some of the programs I saw for estim audio generation this is very simplistic implementation.

**Comment by MattJ-max** [comment]
I've been hacking at this the last few days.  I'm doing it via the MFP plugin system and it's working great.  I'll get my work up on eros scripts soon so some brave folks can hook their business to it and test it out in the next couple of days. 

I'm working on two plugins right now, and might fiddle with a third when the first two are finished.

1.  Stereostim.   Loads an MP3 file, lets you select an audio output from the list of available WASAPI devices, and plays it in sync with whatever video player is linked to MFP.   This is what most people are looking for in the above comments.  I'm adding some basic shaping controls so folks can adjust per-channel volume and a basic eq for tuning up the parts you like and down the parts you don't.   

2.  Coyote DG3 HWL Script player - Loads a HWL file (https://github.com/Amethyst-Sysadmin/Howl) , connects to a DG-Labs Coyote device via bluetooth and plays the file in sync with the video.  I'm also working on an mp3 > HWL converter to create the HWL files from existing stim audio files.   

I've been thinking about another angle and would be interested to know if anyone's excited about it...  it would work with the Coyote like plugin 2 does, but instead of playing back a stim script in the hwl format, it would let you link a normal MFP axis to the coyote's waveform generator so you could make amplitude and frequency follow a fun script's motion.  Honestly, this third idea might make sense to eventually fold into MFP as another output mode since it's already got the framework to hold connection state, assign a script axis to an output axis, set limits on the output axis and all of that...  which is stuff I'd have to duplicate to do this one as a plugin (which is why I haven't yet!)   

If anyone's got thoughts let me know.  I'll post again when it's up on Eros.
